src/cuter_types.erl

```diff
@@ -804,15 +804,124 @@ maybe_hash_args(Ps, Args) ->
   Ps ++ [integer_to_list(Arity), integer_to_list(erlang:phash2(Args))].
 
 %% ----------------------------------------------------------------------------
+%% Normalize parsed specs & type dependencies
+%% ----------------------------------------------------------------------------
+
+normalize_type_deps(Deps) ->
+  Seen = ets:new(?MODULE, [ordered_set, protected]),
+  NormalizedDeps = normalize_type_deps(Deps, Seen, []),
+  ets:delete(Seen),
+  NormalizedDeps.
+
+normalize_type_deps([], _Seen, Acc) ->
+  lists:reverse(Acc);
+normalize_type_deps([{TypeName, Type}|Types], Seen, Acc) ->
+  {NormalizedType, MoreDeps} = normalize_single_type(Type, Seen, true),
+  Acc1 = lists:reverse(MoreDeps) ++ Acc,
```

Comment:
You can avoid the double traversal of the first list here by writing this as `lists:reverse(MoreDeps, Acc)` instead.